### PR TITLE
Add esports notifications

### LINKED ISSUE #101: Post when new esports events go live or are scheduled
Check whether or not the twitch stream is live and/or check the schedule and post the streams when the events will be on twitch/youtube:

https://eu.lolesports.com/en/api/widget/schedule?timezone=Europe%2FAmsterdam&leagues=26&leagues=3&leagues=2&leagues=6&leagues=7&leagues=5&leagues=4&leagues=9&leagues=10&leagues=1&leagues=43&slug=all

### PR DESCRIPTION
**Pull request by stelar7**
Fixes #101

### CODE REVIEW

File: src/ESports.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    teamA: string;
+    teamB: string;
+}
+
+export default class ESportsAPI {
+    private bot: Discord.Client;
+    private settings: SharedSettings;
+
+    // Map<String, Map<String, ESportsLeagueSchedule>>
+    private schedule: { [date: string]: { [league: string]: ESportsLeagueSchedule[] } } = {};
```
Comment (molenzwiebel) [inline src/ESports.ts]: You could always use real maps here ;) Javascript has em.

File: src/ESports.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            await this.loadData();
+            this.postInfo();
+        });
+    }
+
+    public onCheckNext(message: Discord.Message, isAdmin: boolean, command: string, args: string[]) {
+        if (args.length !== 1) return;
+
+        // day / month
+        const formatCheck = new RegExp("\\d\\d/\\d\\d");
```
Comment (molenzwiebel) [inline src/ESports.ts]: `const formatCheck = /\d\d\/\d\d/;`

File: src/ESports.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        if (!formatCheck.test(args[0])) return;
+
+        const data = args[0].split("/");
+        const date = `${data[0]} ${data[1]} ${new Date().getFullYear()}`;
```
Comment (molenzwiebel) [inline src/ESports.ts]: what if the game is next year 🤔 




jk

File: src/ESports.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            console.error(`Esports: Unable to find channel #esports-spoilers`);
+            return;
+        }
+
+        const now = new Date();
+        const date = `${now.getDate()} ${("0" + (now.getMonth() + 1)).slice(-2)} ${now.getFullYear()}`;
+
+        // filter old games
+        const prints: { [league: string]: ESportsLeagueSchedule[] } = {};
+        Object.keys(this.schedule[date]).forEach(league => {
```
Comment (molenzwiebel) [inline src/ESports.ts]: `for (const [league, entry] of Object.entries(this.schedule))` might be cleaner. Or use a map

File: src/ESports.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            prints[league] = this.schedule[date][league].filter(e => e.time.indexOf("ago") === -1);
```
Comment (molenzwiebel) [inline src/ESports.ts]: Maybe `!e.time.includes("ago")`?

File: src/ESports.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        });
+
+        this.sendPrintout(esports as Discord.TextChannel, prints, date);
+        setTimeout(this.postInfo, 3600000);
+    }
+
+    private sendPrintout(channel: Discord.TextChannel, data: { [league: string]: ESportsLeagueSchedule[] }, date: string) {
+        let output = `Games being played ${date.split(" ").join("/")}:\n\`\`\``;
+        const padLeague = Object.keys(data).reduce((a, b) => a.length > b.length ? a : b).length;
+        Object.keys(data).forEach(league => {
```
Comment (molenzwiebel) [inline src/ESports.ts]: `for (const [league, games] of Object.entries(data))`

File: src/ESports.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        // for each date
+        const asHtml = CheerioAPI.load(html);
+        const parents = asHtml.root().find(".schedule__row-group");
+        parents.each((indexed, elem) => {
+            // the current date
+            const elemRoot = CheerioAPI.load(elem).root();
+            const date = elemRoot.find(".schedule__row--date").find("h2").text().split(" ");
+
+            // hack to get month number from text
+            const month = ("0" + ("JanFebMarAprMayJunJulAugSepOctNovDec".indexOf(date[2].substr(0, 3)) / 3 + 1)).slice(-2);
```
Comment (molenzwiebel) [inline src/ESports.ts]: what the fuck

File: src/ESports.ts (hunk 1 above)
Comment (stelar7) [reply src/ESports.ts]: True, its just a remnant of not being able to use maps with the FileBackedObject

File: src/ESports.ts (hunk 3 above)
Comment (stelar7) [reply src/ESports.ts]: 🤔

File: src/ESports.ts (hunk 7 above)
Comment (stelar7) [reply src/ESports.ts]: If you have a better way of turning a string date into a number ill take it 🤷‍♂️

Comment (Querijn) [reply src/ESports.ts]: ```const jsDate = new Date(`${date[1]} ${date[2]}`);
const month = jsDate.getMonth() + 1;```

Comment (Querijn) [reply src/ESports.ts]: Thanks Github for making that a oneliner

File: src/ESports.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    private async loadData() {
+        // pull data
+        const data = await fetch("https://eu.lolesports.com/en/api/widget/schedule?timezone=Europe%2FOslo&leagues=26&leagues=3&leagues=2&leagues=6&leagues=7&leagues=5&leagues=4&leagues=9&leagues=10&leagues=1&leagues=43&slug=all");
+        const html = (await data.json() as ESportsAPIReturnData).fixturesHtml;
+        const schedule: Map<string, Map<string, ESportsLeagueSchedule[]>> = new Map();
+
+        // for each date
+        const asHtml = CheerioAPI.load(html);
+        const parents = asHtml.root().find(".schedule__row-group");
+        parents.each((indexed, elem) => {
```
Comment (Querijn) [inline src/ESports.ts]: Can you type these?

File: src/ESports.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            // the current date
+            const elemRoot = CheerioAPI.load(elem).root();
+            const date = elemRoot.find(".schedule__row--date").find("h2").text().split(" ");
+
+            // hack to get month number from text
+            const month = ("0" + ("JanFebMarAprMayJunJulAugSepOctNovDec".indexOf(date[2].substr(0, 3)) / 3 + 1)).slice(-2);
+
+            const realDate = `${new Date().getFullYear()} ${month} ${date[1]}`;
```
Comment (Querijn) [inline src/ESports.ts]: This might be incorrect if games are played in january while we're in december.

File: src/ESports.ts [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+    private postInfo() {
+        const esports = this.bot.guilds.get(this.settings.server)!.channels.find("name", "esports-spoilers");
+        if (!esports) {
+            console.error(`Esports: Unable to find channel #esports-spoilers`);
+            return;
+        }
+
+        const tomorrow = new Date();
+        tomorrow.setHours(tomorrow.getHours() + 12);
```
Comment (Querijn) [inline src/ESports.ts]: Can you explain why only 12 hours are needed to get to tomorrow? Does this always work?

File: src/ESports.ts (hunk 7 above)
Comment (Querijn) [reply src/ESports.ts]: Heck you could even replace ${date[1]} with 1 and it would work

File: src/ESports.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+                        time: difference,
+                        teamA,
+                        teamB,
+                    };
+                    schedule.get(realDate)!.get(title)!.push(gameData);
+                }
+            }
+        });
+        this.schedule = schedule;
+        setTimeout(this.loadData, 1000 * 60 * 60 * 5); // 5 hours
```
Comment (Querijn) [inline src/ESports.ts]: Can we make this number a setting

File: src/ESports.ts [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        if (!formatCheck.test(args[0])) return;
+
+        const data = args[0].split("/");
+        const date = `${data[0]} ${data[1]} ${data[2]}`;
+
+        this.sendPrintout(message.channel as Discord.TextChannel, this.schedule.get(date), date);
+    }
+
+    private postInfo() {
+        const esports = this.bot.guilds.get(this.settings.server)!.channels.find("name", "esports-spoilers");
```
Comment (Querijn) [inline src/ESports.ts]: Can we make the channel name a setting

File: src/ESports.ts [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        if (!esports) {
+            console.error(`Esports: Unable to find channel #esports-spoilers`);
```
Comment (Querijn) [inline src/ESports.ts]: Don't forget putting the channel name setting in here as well

File: src/ESports.ts [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+                        }
+
+                        prints.get(league)!.push(item);
+                    }
+                }
+            }
+        }
+
+        this.sendPrintout(esports as Discord.TextChannel, prints, tellDate);
+        setTimeout(this.postInfo, 1000 * 60 * 60 * 1); // once per hour
```
Comment (Querijn) [inline src/ESports.ts]: Make this a setting

File: src/ESports.ts (hunk 10 above)
Comment (stelar7) [reply src/ESports.ts]: Should be 24, but i didnt like that it output `in 2 days` from moment.js, so i reduced it 🤷‍♂️

File: src/ESports.ts (hunk 9 above)
Comment (stelar7) [reply src/ESports.ts]: Not sure how that case should be handled 🤔

File: src/ESports.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        const parents = asHtml.root().find(".schedule__row-group");
+        let currentYear = new Date().getFullYear();
+        let currentMonth = new Date().getMonth() + 1;
+        parents.each((_, dayGroup) => {
+            // the current date
+            const dayRoot = CheerioAPI.load(dayGroup).root();
+            const date = dayRoot.find(".schedule__row--date").find("h2").text().split(" ");
+
+            // hack to get month number from text
+            // const month = ("0" + ("JanFebMarAprMayJunJulAugSepOctNovDec".indexOf(date[2].substr(0, 3)) / 3 + 1)).slice(-2);
```
Comment (molenzwiebel) [inline src/ESports.ts]: delet dis

File: src/ESports.ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+            this.postInfo();
+        });
+    }
+
+    public onCheckNext(message: Discord.Message, isAdmin: boolean, command: string, args: string[]) {
+        if (args.length !== 1) return;
+
+        // YYYY/MM/DD
+        const formatCheck = /\d{4}\/\d{2}\/\d{2}/;
+        if (!formatCheck.test(args[0])) return;
```
Comment (molenzwiebel) [inline src/ESports.ts]: Maybe send a message here?

File: src/ESports.ts [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+        });
+    }
+
+    public onCheckNext(message: Discord.Message, isAdmin: boolean, command: string, args: string[]) {
+        if (args.length !== 1) return;
+
+        // YYYY/MM/DD
+        const formatCheck = /\d{4}\/\d{2}\/\d{2}/;
+        if (!formatCheck.test(args[0])) {
+            message.channel.send("The date you specified didnt match the format needed. (YYYY/MM/DD)");
```
Comment (molenzwiebel) [inline src/ESports.ts]: Just add a ‘ to didnt and I’ll merge. Approving for now.

**Pull request merged**